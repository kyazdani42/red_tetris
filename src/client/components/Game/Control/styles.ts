import styled from 'styled-components';

export const KeyStyle: any = styled.div`
  width: 40px;
  height: 40px;
  color: #000;
`;

export const KeyImg: any = styled.img`
  background-color: ${(props: any) => props.isPressed ? '#555' : '#fff'};
  filter: invert(1);
  width: 40px;
`;

export const ControllerWrapper = styled.div`
  display: flex;
  align-items: flex-end;
  flex-direction: row;
  justify-content: center;
`;

export const KeyWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

export const BottomWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
`;
