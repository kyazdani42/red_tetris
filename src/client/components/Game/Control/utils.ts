export const setWindowEvents = (emitAction: (type: keyType | null) => () => void) => {
  const listener = handleKeyPress(emitAction);
  window.removeEventListener('keydown', listener);
  window.addEventListener('keydown', listener);
  return listener;
};

export const handleKeyPress = (emitAction: (type: keyType | null) => () => void) => (e: any) => {
  switch (e.key) {
    case ' ':
      emitAction(' ')();
      break;
    case 'ArrowUp':
      emitAction('up')();
      break;
    case 'ArrowDown':
      emitAction('down')();
      break;
    case 'ArrowLeft':
      emitAction('left')();
      break;
    case 'ArrowRight':
      emitAction('right')();
      break;
  }
};

export const emitter = (
  socket: SocketIOClient.Socket,
  setClick: React.Dispatch<React.SetStateAction<keyType | null>>
) => (type: keyType | null) => () => {
  const emitString = getEmitStringFromType(type);
  socket.emit(emitString);
  setClick(type);
  setTimeout(() => setClick(null), 60);
};

export const getEmitStringFromType = (type: keyType | null) => {
  switch (type) {
    case ' ':
      return 'goDown';
    case 'down':
      return 'moveDown';
    case 'up':
      return 'rotate';
    case 'left':
      return 'moveLeft';
    case 'right':
      return 'moveRight';
    default:
      return '';
  }
};
