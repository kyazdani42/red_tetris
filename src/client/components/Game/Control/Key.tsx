import * as React from 'react';

import { KeyImg, KeyStyle } from './styles';

interface Props {
  emitter: () => void;
  type: keyType;
  keyPressed: keyType | null;
}

const Key: React.SFC<Props> = ({ keyPressed, emitter, type }) => (
  <KeyStyle onClick={emitter}>
    <KeyImg
      className="key-img"
      src={`/assets/icon${type === ' ' ? 'space' : type}.png`}
      isPressed={keyPressed === type}
    />
  </KeyStyle>
);

export default Key;
