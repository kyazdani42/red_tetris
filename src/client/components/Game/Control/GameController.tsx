import * as React from 'react';
import { connect } from 'react-redux';

import Key from './Key';

import { State } from '../../../store';
import { BottomWrapper, ControllerWrapper, KeyWrapper } from './styles';
import { emitter, setWindowEvents } from './utils';

interface Props {
  socket: SocketIOClient.Socket;
}

const GameController: React.SFC<Props> = ({ socket }) => {
  const [click, setClick] = React.useState<keyType | null>(null);
  const emitAction = emitter(socket, setClick);
  React.useEffect(() => {
    const listener = setWindowEvents(emitAction);
    return () => window.removeEventListener('keydown', listener);
  }, []);
  return (
    <ControllerWrapper>
      <KeyWrapper>
        <Key type="up" keyPressed={click} emitter={emitAction('up')} />
        <BottomWrapper>
          <Key type="left" keyPressed={click} emitter={emitAction('left')} />
          <Key type="down" keyPressed={click} emitter={emitAction('down')} />
          <Key type="right" keyPressed={click} emitter={emitAction('right')} />
        </BottomWrapper>
      </KeyWrapper>
      <Key type=" " keyPressed={click} emitter={emitAction(' ')} />
    </ControllerWrapper>
  );
};

const mapStateToProps = (state: State) => ({
  socket: state.app.socket
});

export default connect(mapStateToProps)(GameController);
