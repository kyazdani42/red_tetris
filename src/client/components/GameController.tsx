import * as React from 'react';
import { connect } from 'react-redux';

import Key from './Key';

import { keyDown, keyLeft, keyRight, keySpace, keyUp } from '../actions/actions';

interface Props {
  dispatchKeySpace: () => void;
  dispatchKeyUp: () => void;
  dispatchKeyDown: () => void;
  dispatchKeyLeft: () => void;
  dispatchKeyRight: () => void;
}

const handleKeyPress = (props: Props) => (e: any) => {
  switch (e.key) {
    case ' ':
      props.dispatchKeySpace();
      break;
    case 'ArrowUp':
      props.dispatchKeyUp();
      break;
    case 'ArrowDown':
      props.dispatchKeyDown();
      break;
    case 'ArrowLeft':
      props.dispatchKeyLeft();
      break;
    case 'ArrowRight':
      props.dispatchKeyRight();
      break;
  }
};

const setWindowEvents = (props: Props) => {
  const listener = handleKeyPress(props);
  window.removeEventListener('keydown', listener);
  window.addEventListener('keydown', listener);
};

const GameController = (props: Props) => {
  setWindowEvents(props);
  return (
    <div style={{ display: 'flex', flexDirection: 'row' }}>
      <Key type="left" dispatcher={props.dispatchKeyLeft} />
      <Key type="right" dispatcher={props.dispatchKeyRight} />
      <Key type="down" dispatcher={props.dispatchKeyDown} />
      <Key type="up" dispatcher={props.dispatchKeyUp} />
      <Key type=" " dispatcher={props.dispatchKeySpace} />
    </div>
  );
};

const mapDispatchToProps = (dispatch: any) => ({
  dispatchKeySpace: () => dispatch(keySpace()),
  dispatchKeyUp: () => dispatch(keyUp()),
  dispatchKeyDown: () => dispatch(keyDown()),
  dispatchKeyLeft: () => dispatch(keyLeft()),
  dispatchKeyRight: () => dispatch(keyRight()),
});

export default connect(undefined, mapDispatchToProps)(GameController);
