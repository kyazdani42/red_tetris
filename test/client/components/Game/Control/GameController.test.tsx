// tslint:disable
import * as React from 'react';
import { shallow } from 'enzyme';

import { GameController } from '../../../../../src/client/components/Game/Control/GameController';

describe('testing the component', () => {
  const props = {
    keyPressed: null,
    dispatchHandleKey: jest.fn()
  };
  const wrapper = shallow(<GameController {...props} />);
  it('smoke tests the component', () => {
    expect(wrapper.update()).toMatchSnapshot();
  });
});
