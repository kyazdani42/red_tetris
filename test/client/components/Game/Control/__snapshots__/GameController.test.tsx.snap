// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the component smoke tests the component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    dispatchHandleKey={[Function]}
    keyPressed={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <UNDEFINED>
        <UNDEFINED>
          <Key
            emitter={[Function]}
            keyPressed={null}
            type="left"
          />
          <UNDEFINED>
            left
          </UNDEFINED>
        </UNDEFINED>
        <UNDEFINED>
          <Key
            emitter={[Function]}
            keyPressed={null}
            type="right"
          />
          <UNDEFINED>
            right
          </UNDEFINED>
        </UNDEFINED>
        <UNDEFINED>
          <Key
            emitter={[Function]}
            keyPressed={null}
            type="down"
          />
          <UNDEFINED>
            down
          </UNDEFINED>
        </UNDEFINED>
        <UNDEFINED>
          <Key
            emitter={[Function]}
            keyPressed={null}
            type="up"
          />
          <UNDEFINED>
            rotate
          </UNDEFINED>
        </UNDEFINED>
        <UNDEFINED>
          <Key
            emitter={[Function]}
            keyPressed={null}
            type=" "
          />
          <UNDEFINED>
            drop
          </UNDEFINED>
        </UNDEFINED>
      </UNDEFINED>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <UNDEFINED>
            <Key
              emitter={[Function]}
              keyPressed={null}
              type="left"
            />
            <UNDEFINED>
              left
            </UNDEFINED>
          </UNDEFINED>,
          <UNDEFINED>
            <Key
              emitter={[Function]}
              keyPressed={null}
              type="right"
            />
            <UNDEFINED>
              right
            </UNDEFINED>
          </UNDEFINED>,
          <UNDEFINED>
            <Key
              emitter={[Function]}
              keyPressed={null}
              type="down"
            />
            <UNDEFINED>
              down
            </UNDEFINED>
          </UNDEFINED>,
          <UNDEFINED>
            <Key
              emitter={[Function]}
              keyPressed={null}
              type="up"
            />
            <UNDEFINED>
              rotate
            </UNDEFINED>
          </UNDEFINED>,
          <UNDEFINED>
            <Key
              emitter={[Function]}
              keyPressed={null}
              type=" "
            />
            <UNDEFINED>
              drop
            </UNDEFINED>
          </UNDEFINED>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Key
                emitter={[Function]}
                keyPressed={null}
                type="left"
              />,
              <UNDEFINED>
                left
              </UNDEFINED>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "emitter": [Function],
                "keyPressed": null,
                "type": "left",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "left",
              },
              "ref": null,
              "rendered": "left",
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-EHOje",
                  "isStatic": true,
                  "rules": Array [
                    "
  color: #fff;
",
                  ],
                },
                "displayName": "styled.span",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-EHOje",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-ifAKCX",
              "isStatic": true,
              "rules": Array [
                "
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-ifAKCX",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Key
                emitter={[Function]}
                keyPressed={null}
                type="right"
              />,
              <UNDEFINED>
                right
              </UNDEFINED>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "emitter": [Function],
                "keyPressed": null,
                "type": "right",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "right",
              },
              "ref": null,
              "rendered": "right",
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-EHOje",
                  "isStatic": true,
                  "rules": Array [
                    "
  color: #fff;
",
                  ],
                },
                "displayName": "styled.span",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-EHOje",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-ifAKCX",
              "isStatic": true,
              "rules": Array [
                "
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-ifAKCX",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Key
                emitter={[Function]}
                keyPressed={null}
                type="down"
              />,
              <UNDEFINED>
                down
              </UNDEFINED>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "emitter": [Function],
                "keyPressed": null,
                "type": "down",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "down",
              },
              "ref": null,
              "rendered": "down",
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-EHOje",
                  "isStatic": true,
                  "rules": Array [
                    "
  color: #fff;
",
                  ],
                },
                "displayName": "styled.span",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-EHOje",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-ifAKCX",
              "isStatic": true,
              "rules": Array [
                "
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-ifAKCX",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Key
                emitter={[Function]}
                keyPressed={null}
                type="up"
              />,
              <UNDEFINED>
                rotate
              </UNDEFINED>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "emitter": [Function],
                "keyPressed": null,
                "type": "up",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "rotate",
              },
              "ref": null,
              "rendered": "rotate",
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-EHOje",
                  "isStatic": true,
                  "rules": Array [
                    "
  color: #fff;
",
                  ],
                },
                "displayName": "styled.span",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-EHOje",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-ifAKCX",
              "isStatic": true,
              "rules": Array [
                "
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-ifAKCX",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Key
                emitter={[Function]}
                keyPressed={null}
                type=" "
              />,
              <UNDEFINED>
                drop
              </UNDEFINED>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "emitter": [Function],
                "keyPressed": null,
                "type": " ",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "drop",
              },
              "ref": null,
              "rendered": "drop",
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-EHOje",
                  "isStatic": true,
                  "rules": Array [
                    "
  color: #fff;
",
                  ],
                },
                "displayName": "styled.span",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-EHOje",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-ifAKCX",
              "isStatic": true,
              "rules": Array [
                "
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-ifAKCX",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bxivhb",
          "isStatic": true,
          "rules": Array [
            "
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  align-items: center;
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bxivhb",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-htpNat",
        "isStatic": true,
        "rules": Array [
          "
  height: 100%;
  width: 100%;
",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-htpNat",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <UNDEFINED>
          <UNDEFINED>
            <Key
              emitter={[Function]}
              keyPressed={null}
              type="left"
            />
            <UNDEFINED>
              left
            </UNDEFINED>
          </UNDEFINED>
          <UNDEFINED>
            <Key
              emitter={[Function]}
              keyPressed={null}
              type="right"
            />
            <UNDEFINED>
              right
            </UNDEFINED>
          </UNDEFINED>
          <UNDEFINED>
            <Key
              emitter={[Function]}
              keyPressed={null}
              type="down"
            />
            <UNDEFINED>
              down
            </UNDEFINED>
          </UNDEFINED>
          <UNDEFINED>
            <Key
              emitter={[Function]}
              keyPressed={null}
              type="up"
            />
            <UNDEFINED>
              rotate
            </UNDEFINED>
          </UNDEFINED>
          <UNDEFINED>
            <Key
              emitter={[Function]}
              keyPressed={null}
              type=" "
            />
            <UNDEFINED>
              drop
            </UNDEFINED>
          </UNDEFINED>
        </UNDEFINED>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <UNDEFINED>
              <Key
                emitter={[Function]}
                keyPressed={null}
                type="left"
              />
              <UNDEFINED>
                left
              </UNDEFINED>
            </UNDEFINED>,
            <UNDEFINED>
              <Key
                emitter={[Function]}
                keyPressed={null}
                type="right"
              />
              <UNDEFINED>
                right
              </UNDEFINED>
            </UNDEFINED>,
            <UNDEFINED>
              <Key
                emitter={[Function]}
                keyPressed={null}
                type="down"
              />
              <UNDEFINED>
                down
              </UNDEFINED>
            </UNDEFINED>,
            <UNDEFINED>
              <Key
                emitter={[Function]}
                keyPressed={null}
                type="up"
              />
              <UNDEFINED>
                rotate
              </UNDEFINED>
            </UNDEFINED>,
            <UNDEFINED>
              <Key
                emitter={[Function]}
                keyPressed={null}
                type=" "
              />
              <UNDEFINED>
                drop
              </UNDEFINED>
            </UNDEFINED>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <Key
                  emitter={[Function]}
                  keyPressed={null}
                  type="left"
                />,
                <UNDEFINED>
                  left
                </UNDEFINED>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "emitter": [Function],
                  "keyPressed": null,
                  "type": "left",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": "left",
                },
                "ref": null,
                "rendered": "left",
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-EHOje",
                    "isStatic": true,
                    "rules": Array [
                      "
  color: #fff;
",
                    ],
                  },
                  "displayName": "styled.span",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-EHOje",
                  "target": "span",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                },
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-ifAKCX",
                "isStatic": true,
                "rules": Array [
                  "
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-ifAKCX",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <Key
                  emitter={[Function]}
                  keyPressed={null}
                  type="right"
                />,
                <UNDEFINED>
                  right
                </UNDEFINED>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "emitter": [Function],
                  "keyPressed": null,
                  "type": "right",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": "right",
                },
                "ref": null,
                "rendered": "right",
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-EHOje",
                    "isStatic": true,
                    "rules": Array [
                      "
  color: #fff;
",
                    ],
                  },
                  "displayName": "styled.span",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-EHOje",
                  "target": "span",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                },
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-ifAKCX",
                "isStatic": true,
                "rules": Array [
                  "
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-ifAKCX",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <Key
                  emitter={[Function]}
                  keyPressed={null}
                  type="down"
                />,
                <UNDEFINED>
                  down
                </UNDEFINED>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "emitter": [Function],
                  "keyPressed": null,
                  "type": "down",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": "down",
                },
                "ref": null,
                "rendered": "down",
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-EHOje",
                    "isStatic": true,
                    "rules": Array [
                      "
  color: #fff;
",
                    ],
                  },
                  "displayName": "styled.span",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-EHOje",
                  "target": "span",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                },
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-ifAKCX",
                "isStatic": true,
                "rules": Array [
                  "
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-ifAKCX",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <Key
                  emitter={[Function]}
                  keyPressed={null}
                  type="up"
                />,
                <UNDEFINED>
                  rotate
                </UNDEFINED>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "emitter": [Function],
                  "keyPressed": null,
                  "type": "up",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": "rotate",
                },
                "ref": null,
                "rendered": "rotate",
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-EHOje",
                    "isStatic": true,
                    "rules": Array [
                      "
  color: #fff;
",
                    ],
                  },
                  "displayName": "styled.span",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-EHOje",
                  "target": "span",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                },
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-ifAKCX",
                "isStatic": true,
                "rules": Array [
                  "
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-ifAKCX",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <Key
                  emitter={[Function]}
                  keyPressed={null}
                  type=" "
                />,
                <UNDEFINED>
                  drop
                </UNDEFINED>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "emitter": [Function],
                  "keyPressed": null,
                  "type": " ",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": "drop",
                },
                "ref": null,
                "rendered": "drop",
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-EHOje",
                    "isStatic": true,
                    "rules": Array [
                      "
  color: #fff;
",
                    ],
                  },
                  "displayName": "styled.span",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-EHOje",
                  "target": "span",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                },
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-ifAKCX",
                "isStatic": true,
                "rules": Array [
                  "
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-ifAKCX",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bxivhb",
            "isStatic": true,
            "rules": Array [
              "
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  align-items: center;
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bxivhb",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-htpNat",
          "isStatic": true,
          "rules": Array [
            "
  height: 100%;
  width: 100%;
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-htpNat",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
